### PR TITLE
fix(browser): only import file I/O functions when running in Node.js environments

### PR DESCRIPTION
**Pull request by krishna-acondy**
## Issue

Fixes https://github.com/sasjs/adapter/issues/493

## Intent

Fix the adapter build to work both in the browser and in Node.js environments.

## Implementation
* Added `isNode` utility function.
* Separated creation of file streams into its own module `getFileStream`.
* Used `require` statements to dynamically import certain functions if running in Node.js environment.
* Fixed imports in `getTokens` module - now importing from separate barrels rather than the main one.
* Fixed and added tests as necessary.

## Checks

No PR (that involves a non-trivial code change) should be merged, unless all items below are confirmed!  If an urgent fix is needed - use a tar file.

### Deployed Apps
✅ [React - Viya Web](https://sas.analytium.co.uk/kriaco/react-web)
✅ [React - Viya JES](https://sas.analytium.co.uk/kriaco/react-jes)
✅ [React - Viya Compute](https://sas.analytium.co.uk/kriaco/react-compute)
✅ [React - SAS9](https://sas.analytium.co.uk:8343/krishna/react-sas9)

I've not deployed the Angular Seed App since testing Data Controller covers that.

✅ [Data Controller - Viya Web](https://sas.analytium.co.uk/kriaco/dc-web)
✅ [Data Controller - Viya JES](https://sas.analytium.co.uk/kriaco/dc-jes)
✅ [Data Controller - Viya Compute](https://sas.analytium.co.uk/kriaco/dc-compute)
✅ [Data Controller - SAS9](https://sas.analytium.co.uk:8343/krishna/dc-sas9)

✅ [SASjs Tests - Viya Web](https://sas.analytium.co.uk/kriaco/sasjs-tests-web)
✅ [SASjs Tests - Viya JES](https://sas.analytium.co.uk/kriaco/sasjs-tests-jes)
✅ [SASjs Tests - Viya Compute](https://sas.analytium.co.uk/kriaco/sasjs-tests-compute)
🟡 [SASjs Tests - SAS9](https://sas.analytium.co.uk:8343/krishna/sasjs-tests-sas9) - known issues being fixed in https://github.com/sasjs/adapter/pull/487

- [x] All `sasjs-cli` unit tests are passing (`npm test`).
### Mocked Tests
image

### Server Tests
image

- [x] All `sasjs-tests` are passing (instructions available [here](https://github.com/sasjs/adapter/blob/master/sasjs-tests/README.md)).
- [x] [Data Controller](https://datacontroller.io) builds and is functional on both SAS 9 and Viya

### CODE REVIEW

**Review by medjedovicm — APPROVED** [review]

File: src/api/viya/getFileStream.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { isFolder } from '@sasjs/utils/file'
+import { generateTimestamp } from '@sasjs/utils/time'
+import { Job } from '../../types'
+
+export const getFileStream = async (job: Job, filePath?: string) => {
```
Comment (krishna-acondy) [inline src/api/viya/getFileStream.ts:5]: Moving this functionality into a separate file lets us remove the `fs` import in `pollJobState.ts`.
The `require` on line 6 also makes sure we only import the `createWriteStream` function when it is actually needed.

File: src/api/viya/pollJobState.ts [hunk 2]
```diff
@@ -55,21 +52,9 @@ export async function pollJobState(
...
-      const logFilePath = `${
-        pollOptions?.logFolderPath || process.cwd()
-      }/${logFileName}`
-
-      logFileStream = await createWriteStream(logFilePath)
-    } else {
-      logFileStream = await createWriteStream(logPath)
-    }
+  if (pollOptions.streamLog && isNode()) {
+    const { getFileStream } = require('./getFileStream')
```
Comment (krishna-acondy) [inline src/api/viya/pollJobState.ts:56]: Only create a write stream if `isNode()` is `true`.

File: src/api/viya/pollJobState.ts [hunk 3]
```diff
@@ -230,14 +215,17 @@ const doPoll = async (
...
-      await saveLog(
-        postedJob,
-        requestClient,
-        startLogLine,
-        endLogLine,
-        logStream,
-        authConfig?.access_token
-      )
+      const { saveLog } = isNode() ? require('./saveLog') : { saveLog: null }
+      if (saveLog) {
```
Comment (krishna-acondy) [inline src/api/viya/pollJobState.ts:219]: Only attempt to save to log if `isNode()` is `true`.

File: src/types/WriteStream.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export interface WriteStream {
+  write: (content: string, callback: (err?: Error) => any) => void
```
Comment (krishna-acondy) [inline src/types/WriteStream.ts:2]: Replaced the import of `fs.WriteStream` with our own, locally defined interface.

File: src/api/viya/getFileStream.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  const { createWriteStream } = require('@sasjs/utils/file')
+  const logPath = filePath || process.cwd()
+  const isFolderPath = await isFolder(logPath)
+  if (isFolderPath) {
+    const logFileName = `${job.name || 'job'}-${generateTimestamp()}.log`
+    const logFilePath = `${filePath || process.cwd()}/${logFileName}`
```
Comment (allanbowe) [inline src/api/viya/getFileStream.ts:11]: does the `/` work on windows?

Comment (krishna-acondy) [reply src/api/viya/getFileStream.ts]: No, but it should now.

**Review by saadjutt01 — DISMISSED** [review]

File: src/api/viya/spec/getFileStream.spec.ts [hunk 6]
```diff
@@ -27,7 +27,7 @@ describe('getFileStream', () => {
 
     expect(createWriteStream).not.toHaveBeenCalledWith(__dirname)
     expect(createWriteStream).toHaveBeenCalledWith(
-      expect.stringContaining(__dirname + '/test job-20')
+      expect.stringContaining(path.join(__dirname, '/test job-20'))
```
Comment (allanbowe) [inline src/api/viya/spec/getFileStream.spec.ts:30]: I guess the `/` should be removed from this line?  else I'm terribly mistaken

Comment (krishna-acondy) [reply src/api/viya/spec/getFileStream.spec.ts]: Spot on.

Comment (krishna-acondy) [reply src/api/viya/spec/getFileStream.spec.ts]: Fixed.

**Review by allanbowe — APPROVED** [review]

**Review by medjedovicm — APPROVED** [review]

**Pull request merged**